flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java

```diff
@@ -276,16 +284,21 @@ private void assignPendingPartitionSplits() {
                             .addAll(newPartitionSplits);
                     // Clear the pending splits for the reader owner.
                     pendingPartitionSplitAssignment.remove(readerOwner);
-                    // Sends NoMoreSplitsEvent to the readers if there is no more partition splits
-                    // to be assigned.
-                    if (noMoreNewPartitionSplits) {
-                        LOG.debug(
-                                "No more KafkaPartitionSplits to assign. Sending NoMoreSplitsEvent to the readers "
-                                        + "in consumer group {}.",
-                                consumerGroupId);
-                        context.signalNoMoreSplits(readerOwner);
-                    }
                 });
+        if (noMoreNewPartitionSplits) {
+            signalNoMoreSplitsToNotNotifiedReaders();
+        }
+    }
+
+    private void signalNoMoreSplitsToNotNotifiedReaders() {
+        context.registeredReaders()
+                .forEach(
+                        (readerId, ignore) -> {
+                            if (!finishedReaders.contains(readerId)) {
+                                context.signalNoMoreSplits(readerId);
+                                finishedReaders.add(readerId);
```

Comment:
What happens if the readers failover? Do we assume that the source readers remember the reception of `NoMoreSplitsEvent`?